### PR TITLE
Replace python random with torch.rand to enable dynamo.export

### LINKED ISSUE #102794: [dynamo.export] `AssertionError: Dynamo attempts to add additional input during export` for Huggingface Bart
### 🐛 Describe the bug

Model source code of root cause is
```python
            # add LayerDrop (see https://arxiv.org/abs/1909.11556 for description)
            dropout_probability = random.uniform(0, 1)
            if self.training and (dropout_probability < self.layerdrop):  # skip the layer
                layer_outputs = (None, None)
```

A naive fix for inference would be changing model source code to inline and short circuit `random.uniform(0, 1)` in the condition.
Any ideas for more fundamental fix are welcomed.

Repro script:

```python
import torch
import torch._dynamo

from transformers import AutoTokenizer, BartForCausalLM

tokenizer = AutoTokenizer.from_pretrained("facebook/bart-base")
model = BartForCausalLM.from_pretrained("facebook/bart-base", add_cross_attention=False)
assert model.config.is_decoder, f"{model.__class__} has to be configured as a decoder."
inputs = tokenizer("Hello, my dog is cute", return_tensors="pt")
outputs = model(**inputs)

logits = outputs.logits
expected_shape = [1, inputs.input_ids.shape[-1], model.config.vocab_size]
list(logits.shape) == expected_shape

torch._dynamo.export(model, **inputs)
```

Log
```
Traceback (most recent call last):
  File "repro_dynamo_insert_additional_input.py", line 16, in <module>
    torch._dynamo.export(model, **inputs)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/eval_frame.py", line 947, in export
    result_traced = opt_f(*args, **kwargs)
  File "/home/bowbao/pytorch_dev/torch/nn/modules/module.py", line 1502, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/bowbao/pytorch_dev/torch/nn/modules/module.py", line 1511, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/eval_frame.py", line 290, in _fn
    return fn(*args, **kwargs)
  File "/home/bowbao/pytorch_dev/torch/nn/modules/module.py", line 1502, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/bowbao/pytorch_dev/torch/nn/modules/module.py", line 1511, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/eval_frame.py", line 443, in catch_errors
    return callback(frame, cache_size, hooks, frame_state)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/convert_frame.py", line 127, in _fn
    return fn(*args, **kwargs)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/convert_frame.py", line 360, in _convert_frame_assert
    return _compile(
  File "/home/bowbao/pytorch_dev/torch/_dynamo/utils.py", line 180, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/convert_frame.py", line 430, in _compile
    out_code = transform_code_object(code, transform)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/bytecode_transformation.py", line 1000, in transform_code_object
    transformations(instructions, code_options)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/convert_frame.py", line 415, in transform
    tracer.run()
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 2020, in run
    super().run()
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 704, in run
    and self.step()
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 664, in step
    getattr(self, inst.opname)(inst)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 386, in wrapper
    return inner_fn(self, inst)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 1147, in CALL_FUNCTION_KW
    self.call_function(fn, args, kwargs)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 555, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/home/bowbao/pytorch_dev/torch/_dynamo/variables/nn_module.py", line 333, in call_function
    return tx.inline_user_function_return(
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 591, in inline_user_function_return
    result = InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 2125, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 2222, in inline_call_
    tracer.run()
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 704, in run
    and self.step()
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 664, in step
    getattr(self, inst.opname)(inst)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 386, in wrapper
    return inner_fn(self, inst)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 1135, in CALL_FUNCTION_EX
    self.call_function(fn, argsvars.items, kwargsvars.items)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 555, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/home/bowbao/pytorch_dev/torch/_dynamo/variables/functions.py", line 333, in call_function
    return super().call_function(tx, args, kwargs)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/variables/functions.py", line 287, in call_function
    return super().call_function(tx, args, kwargs)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/variables/functions.py", line 120, in call_function
    return tx.inline_user_function_return(
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 591, in inline_user_function_return
    result = InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 2125, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 2222, in inline_call_
    tracer.run()
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 704, in run
    and self.step()
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 664, in step
    getattr(self, inst.opname)(inst)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 386, in wrapper
    return inner_fn(self, inst)
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 1095, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "/home/bowbao/pytorch_dev/torch/_dynamo/symbolic_convert.py", line 555, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/home/bowbao/pytorch_dev/torch/_dynamo/variables/user_defined.py", line 302, in call_function
    return VariableBuilder(tx, source).wrap_unspecialized_primitive(
  File "/home/bowbao/pytorch_dev/torch/_dynamo/variables/builder.py", line 988, in wrap_unspecialized_primitive
    raise AssertionError(
AssertionError: Dynamo attempts to add additional input during export: value=0.8042905926704407, source=RandomValueSource(random_call_index=0)

from user code:
   File "/home/bowbao/pytorch_dev/torch/nn/modules/module.py", line 1511, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/bowbao/transformers/src/transformers/models/bart/modeling_bart.py", line 1074, in forward
    dropout_probability = random.uniform(0, 1)

Set torch._dynamo.config.verbose=True or TORCHDYNAMO_VERBOSE=1 for more information


You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True
```

### Versions

From source b02f48b18152ddfcf5fcbefb68f6b66a6c44b37f
transformers                  4.29.2

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy

### PR DESCRIPTION
**Pull request by BowenBao**
# What does this PR do?

Related and Fixes https://github.com/pytorch/pytorch/issues/102794

TL;DR dynamo graph breaks on python `random.uniform(0, 1)`. The graph break can be prevented by replacing with `torch.randn([])`.

Example repro script
```python
import torch
import torch._dynamo

from transformers import AutoTokenizer, BartForCausalLM

tokenizer = AutoTokenizer.from_pretrained("facebook/bart-base")
model = BartForCausalLM.from_pretrained("facebook/bart-base", add_cross_attention=False)
inputs = tokenizer("Hello, my dog is cute", return_tensors="pt")
outputs = model(**inputs)

torch._dynamo.export(model, return_dict=False, **inputs)
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by sgugger — COMMENTED** [review]
You are touching multiple Flax model which shouldn't depend on torch, could you revert that?

**Comment by BowenBao** [comment]
> You are touching multiple Flax model which shouldn't depend on torch, could you revert that?

Nice catch! Done.

**Review by sgugger — APPROVED** [review]
Thanks! @amyeroberts could you have a quick second look here?

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing this! 

Changes LGTM 👍  I think _technically_ this is slightly different as random.uniform can return 1, but for all practical purposes it doesn't change things.

**Comment by anijain2305** [comment]
@BowenBao Does this really solve the export problem. We are seeing export issues here - https://github.com/pytorch/pytorch/issues/107587

If one peeks at the tensor value for the conditional, its a legit dynamic control flow. We might have to use `torch.where` or `torch.cond`.

**Comment by BowenBao** [comment]
@anijain2305 it does for inference export, since the actual condition is short circuited by `self.training` being False. Without the change, `random.uniform(0, 1)` leads to a graph break, although the value is unused.
